Author reconstruction for 'GitHub' user, handling of new GitHub events, handling of GitHub Copilot and other agents

Leo-Send opened the pull request:
This pr adresses issue #52 

Fixes this issue by adding the correct author to field 'event_info_2' in the issue data. This allows for reconstruction of the commit's author during author postprocessing, if the user that envoked the event is the 'GitHub' user.

Since parts of the changes take place in functions relying on codeface, testing all the changes is currently not possible.

Leo-Send committed:
Add commit author of 'commit_added' events to event info

This allows for reconstruction of correct commit author if user is
github

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -178,20 +178,24 @@ def is_github_noreply_author(name, email):
                     # extract commit hash from event info 1
                     commit_hash = event[12]
-
+                    name = event[13][1:-1]
```
Could you please comment on what's going on here? `event[13]` seems to be event.info.2. But what is the purpose of  `[1:-1]`?

Leo-Send replied:
I added a comment above: 
```
# extract author name from event info 2 while cutting excess '"'
```

bockthom reviewed (commented):
Thanks for this PR @Leo-Send.
I don't have run your changes yet, but here are already two comments:

Could you please update your copyright header in the files you have changed?

Leo-Send committed:
Update Copyright headers

also added one comment for clarity

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add connected events reconstruction

also save merge commits
reconstruction of connected events is done by first saving all connected
events that occured at the same time. Then, it is possible to match
connected events iff:
- half of the involved issues are equal, meaning that one issue is
  connected to multiple others
- half rounded up of the involved isses are equal, meaning that we have
  one external connected event and then the previous case with the
remaining issues

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -501,6 +528,10 @@ def merge_issue_events(issue_data):
 
+    # filter out connected events which cannot be perfectly matched
+    global filtered_connected_events
+    filtered_connected_events = dict(filter(lambda item: filter_connected_events(item[0], item[1]), connected_events.iteritems()))
```
Hm, could you please add a reason what "cannot be perfectly matched" means? Otherwise, if someone will be looking at the code later, it may be unclear what happens here.

In addition, I am unhappy about the `global` variable here. Do we have any other global variables in this script? If not, would there by a solution without using a global variable here?

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -490,6 +495,28 @@ def merge_issue_events(issue_data):
 
+            # if event is merged event, save the hash of the merge commit in event_info_1
+            if event["event"] == "merged":
+                event["event_info_1"] = event["commit"]["hash"]
+
+            # if event is connected event, create or add to a matching dict entry by matching timestamps, for later reconstruction
+            if event["event"] == "connected":
+                if event["created_at"] in connected_events.keys() and connected_events[event["created_at"]]["user"] == event["user"]:
+                    connected_events[event["created_at"]]["issues"].append(issue["number"])
+                elif subtract_seconds_from_time(event["created_at"], 1) in connected_events.keys() \
+                        and connected_events[subtract_seconds_from_time(event["created_at"], 1)]["user"] == event["user"]:
+                    connected_events[subtract_seconds_from_time(event["created_at"], 1)]["issues"].append(issue["number"])
+                    event["created_at"] = subtract_seconds_from_time(event["created_at"], 1)
+                elif subtract_seconds_from_time(event["created_at"], -1) in connected_events.keys() \
+                        and connected_events[subtract_seconds_from_time(event["created_at"], -1)]["user"] == event["user"]:
+                    connected_events[subtract_seconds_from_time(event["created_at"], -1)]["issues"].append(issue["number"])
+                    event["created_at"] = subtract_seconds_from_time(event["created_at"], -1)
+                else:
+                    connected_info = dict()
+                    connected_info["issues"] = [issue["number"]]
+                    connected_info["user"] = issue["user"]
+                    connected_events[event["created_at"]] = connected_info
```
Could you please add a short comment for each case for better clarity here?

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -510,6 +541,41 @@ def merge_issue_events(issue_data):
 
+def filter_connected_events(key, value):
+    num_issues = len(value["issues"])
+    global external_connected_events
```
Another global variable here defined within a function. Might be confusing...

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -510,6 +541,41 @@ def merge_issue_events(issue_data):
+        value["multi_issues_copy"] = list(value["issues"])
+        return True
+    # if it is an odd number, check if it can be easily matched
+    # meaning that exactly half (rounded up) the events occur in the same issue
```
Not sure why we need to round up here. What happens to the "one off" event then? Maybe needs some more clarification...

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -540,6 +606,37 @@ def reformat_events(issue_data):
+                            value["issues"].remove(issue["number"])
+                # if so, skip the next checks
+                if external:
+                    continue
```
Does the alignment of for loop and `if external` make sense here? Why do we have to loop over key value pairs but have only one `if external` check afterwards?

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -540,6 +606,37 @@ def reformat_events(issue_data):
+                                    event["event_info_1"] = number
+                                else:
+                                    # if current issue is not the centerpiece, connect it to the centerpiece
+                                    event["event_info_1"] = max(occurances, key = occurances.get)
```
What is the centerpiece? 
Maybe it could be helpful to add real examples (i.e., scenarios that demonstrate how issues are linked) as a multi-line comment...

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -53,6 +54,9 @@
 datetime_format = "%Y-%m-%d %H:%M:%S"
 
+filtered_connected_events = dict()
+external_connected_events = dict()
```
I have an understanding of what external connected events are. But what are filtered connected events?

Leo-Send committed:
Remove unnecessary returns of issue data

since data is modified in-place, return of input data is not needed

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add reasons to reopen/closed events

ALso add commit hash if closed by commit

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add GitHub issue types

also rename 'new feature' to 'feature'

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Simplify loops for reconstruction of connections

also remove duplicates from type list

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add subissues to results csv

using empty line reserved for jira components

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -654,13 +654,16 @@ def reformat_events(issue_data):
             if event["event"] == "closed":
                 event["event"] = "state_updated"
                 event["event_info_1"] = "closed"  # new state
-                event["event_info_2"] = "open"  # old state
+                if event["commit"] is not None:
+                    event["event_info_2"] = event["commit"]["hash"]
+                else:
+                    event["event_info_2"] = event["state_reason"]
                 issue["state_new"] = "closed"
 
             elif event["event"] == "reopened":
                 event["event"] = "state_updated"
                 event["event_info_1"] = "open"  # new state
-                event["event_info_2"] = "closed"  # old state
+                event["event_info_2"] = event["state_reason"]
```
I am not quite sure about this change here. What does the state reason say in this situation? What are the potential reasons for reopening an issue?

Leo-Send replied:
Currently, the state_reason field will always say 'reopened', as it is not possible to provide a reason for reopening issues. We could instead hardcode this instead of just passing the given reason here.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -674,7 +659,7 @@ def reformat_events(issue_data):
                 event["event_info_1"] = label
 
                 # if the label is in this list, it also is a type of the issue
-                if label in known_types:
+                if label in known_types and label not in issue["type"]:
```
Is this something we would also need in the jira processing? Please check.

Leo-Send replied:
While I am not familiar with jira, it is my understanding that labels are a property of the issues there and should therefore always be a unique list.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -611,34 +607,23 @@ def reformat_events(issue_data):
+                            value["multi_issues_copy"].remove(number)
+                            event["event_info_1"] = number
+                        else:
+                            event["event_info_1"] = max(occurances, key = occurances.get)
```
Could you please  add comments here? (similar to the comments you had previously added) Otherwise the if-else cascade is very difficult to understand...

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -79,13 +79,13 @@ def run():
     # 1) load the list of issues
     issues = load(__srcdir)
     # 2) re-format the issues
-    issues = reformat_issues(issues)
+    reformat_issues(issues)
```
Can you please apply this change (and all the related ones below) also to the corresponding functions in jira_issue_processing?

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -81,9 +78,10 @@ def run():
     reformat_issues(issues)
     # 3) merges all issue events into one list
-    merge_issue_events(issues)
+    external_connected_events = dict()
+    filtered_connected_events = merge_issue_events(issues, external_connected_events)
```
Still, I miss explanations for these two variables. Either here or further down where they are used. But some documentation would be helpful...

bockthom reviewed (changes requested):
Could you please address the following issues, as well as the issues previously pointed out by Copilot, as well as the other unresolved one regarding necessary comments from a previous review of mine?

Thank you!

bockthom commented:
And the copyright header in `issue_processing/jira_issue_processing.py` needs to be updated as well.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -538,6 +605,26 @@ def reformat_events(issue_data):
+                            value["multi_issues_copy"].remove(number)
+                            event["event_info_1"] = number
+                        else:
+                            event["event_info_1"] = max(occurances, key = occurances.get)
```
Check how we did this in other function calls. If we don't ever do this, ignore it. If we stick to this kind of style (almost) everywhere else, then please fix it.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -499,16 +529,53 @@ def merge_issue_events(issue_data):
+    # if it is an odd number, check if it can be easily matched
+    # meaning that exactly half (rounded up) the events occur in the same issue
+    if num_issues % 2 == 1 and math.ceil(num_issues/2) in occurances.values():
+        for sub_key, sub_value in occurances.iteritems():
```
True

Leo-Send committed:
Remove unneccesary return value

also added copyright header

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add comments

also minor fixes and removal of math.ceil

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add new json field for suggestions to result

comments now each have a boolean field that describes whether the
comment contains a suggestion or not

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Improve documentation

dicts for reconstructing connected events are now better explained and
the comments do not disruot the workflow in the run function anymore

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -742,7 +742,10 @@ def reformat_events(issue_data, filtered_connected_events, external_connected_ev
+                if "contains_suggestion" in event:
+                    event["event_info_2"] = event["contains_suggestion"]
+                else:
+                    event["event_info_2"] = False
```
The comment does not align with the code any more, please adjust accordingly.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -805,7 +921,7 @@ def print_to_disk(issues, results_folder):
                 issue["created_at"],
                 issue["closed_at"],
-                json.dumps([]),  # components
+                json.dumps([issue["subIssues"]]),  # components
```
Not sure what copilot wants to tell us here. Is this again a python2/python3 conversion issue, or is this an actual problem we need to care about?

Leo-Send replied:
Copilot is correct iff the issue data does not contain the field "subissues", for example if you run this new script on old data extracted using the GitHubWrapper. To be safe, I will initialize the field with an empty list if it is not present.

Leo-Send committed:
Incorporate requested changes

includes:
- updated comments
- spelling mistake
- fix for potential crash if script is used on old data

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add copilot user unification to author postprocessing

author postprocessing now also contains a list of known copilot use
names that can be extended to unify more different copilot users

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Assign copilot user data in case of specific events

the events 'copilot_work_started' and 'copilot_work_finished' now always
have the standard copilot user data

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add documentation for new copilot user unification

Method doc updated to reflect new functionality

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Fix connected event assignment

previously, the creator of the issues was falsely matched to the
connected event instead of the user triggering the event

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -51,6 +51,15 @@
+known_copilot_users = {"Copilot", "copilot-pull-request-reviewer[bot]", "copilot-swe-agentbot"}
+copilot_unified_name = "Copilot"
+copilot_unified_email = "copilot@example.com"
+
```
Can't we move this to a common file? At the moment, the "Copilot" user is used twice: Here and in issue_processing.py.

We could create a new directory called "GitHub_user_utils", and a file "GitHub_user_utils.py" where we store this information. Then we could just import these global variables (those that are actually used in a script) in each of the scripts:

`from GitHub_user_utils import known_copilot_users` (in issue_processing)
`from GitHub_user_utils import known_copilot_users, copilot_unified_name, copilot_unified_email''` (in author_postprocessing)

Maybe we can also move the `github_user` and `github_email` constants (see line 111 below here in this file) and the helper function `is_github_noreply_author` there, not to have different user information distributed among different files.

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -178,20 +189,30 @@ def is_github_noreply_author(name, email):
             issue_data_new = []
 
             for event in issue_data:
+                # unify events to use a single copilot user for all events triggered by a known copilot user
+                if unify_copilot_users and event[9] in known_copilot_users:
+                    event[9] = copilot_unified_name
+                    event[10] = copilot_unified_email
```
(1) Can we add a log statement here to state that we unify the copilot users? Similar to the ones removing GitHub user?

(2) Is this the only place where copilot users can occur? Shouldn't we do this to the authors_list and commits_list as well when copilot creates commits? Not sure about the other data sources as well, please check.

bockthom replied:
This is not enough! 

(i) Think about "commit_added" events, for which the "copilot-swe-agentbot" is part of column 13. [Maybe this is a general oversight: Did we handle this case properly in both author post processing and issue processing? I guess this is newly introduced in this PR, so we should check this.]

(ii) Think about "mentioned" or "subscribed" events, for which the "copilot-swe-agentbot" is part of columns 12 (name) and 13 (email)

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -50,6 +51,15 @@
+##
+
+# global variable containing all known copilot users and the name and mail adress copilot users will be assigned
+known_copilot_users = {"Copilot", "copilot-pull-request-reviewer[bot]", "copilot-swe-agentbot"}
```
This is not sufficient: The ID service may remove special characters such as [ or ]. Thus, we might not be able to merge them correctly.

However, I also want this to be as independent as possible from the ID service., which is why I want to keep the current spelling variant that contains  [ and ].

Suggestion 1: Add both variants: "copilot-pull-request-reviewer[bot]", "copilot-pull-request-reviewerbot"

Suggestion 2: As these square brackets are commonly used, we might only add it once to the list and then add a function that automatically generates the variants without [ and ] in addition. Then `known_copilot_users` should be the outcome of this function.

bockthom replied:
We also need to perform a similar fix in bots processing here: https://github.com/se-sic/codeface-extraction/blob/b4fc74ee49ad2903113f0994a7325a4d3b2b5ba8/bot_processing/bot_processing.py#L195-L199

Here we don't consider the "bot" prefix well: user[0] is "copilot-swe-agent" or "codecov", but the corresponding key in user buffer is "copilot-swe-agentbot" or "codecovbot", respectively. So, we should check if user[0] or user[0] + "bot", or user[0] + "[bot]" is in the the user buffer...

Leo-Send committed:
Unify copilot users in all files

unification now done on all files, which should prevent any issues
arising from unknown authors during anonymization
also move all global variables to a new utils file

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add support for 'known agents'

Known agentsc such as 'copilot' or 'claude' can now be read, similar to
known bots. They will be flagged as agents during bot processing.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add better bot name variant support

Add a helper function for creating bot name variants utilizing either
'[bot]' or 'bot' suffix. Also update bot processing to check user buffer
for all variants.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Add better bot name handling

Add a helper function that given a botname and a list of names, returns
which bot name variant is contained in the list (or None). This is used
whenever we check if a known bot is in the userdata or has been
predicted to be a bot, and means that botnames in the known_bots file do
not need to be duplicated for each variant.
Also, automatically add all known coplilot users to the known_agents
list, and then unify those during author postprocessing.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -50,16 +50,11 @@
+from github_user_utils import known_copilot_users, copilot_unified_name, copilot_unified_email, \
+                              is_github_noreply_author, github_user, github_email, \
+                              commit_added_event, mentioned_event, subscribed_event, generate_botname_variants
 
```
This kind of import does not work. Instead you need 
```python
from github_user_utils.github_user_utils import ...
```

Also affects `bot_processing.py` and `issue_processing.py`.

bockthom commented on github_user_utils/github_user_utils.py:
There is an empty file `__init__.py` missing in folder "github_user_utils". Please create such a file (you can find similar ones in every other folder as well).

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -356,6 +394,9 @@ def run_postprocessing(conf, resdir, backup_data):
                         issue_event[13] = quot_m + person[2] + quot_m
+                    # replace name in event info 2 if necessary
+                    if person[4] == issue_event[13]:
+                        issue_event[13] = person[1]
```
I guess here the `quot_m` is missing around `person[1]`, as column 13 contains these mysterious additional quotes. Both in the check and in the new value.

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -172,26 +181,44 @@ def is_github_noreply_author(name, email):
+                    elif event[8] in (mentioned_event, subscribed_event) and event[12][-1:1] in known_copilot_users_extended:
+                        # for mentioned/subscribed events, also unify the referenced user in event info 1 and 2 if it is a known copilot user
+                        event[12] = '"' + copilot_unified_name + '"'
+                        event[13] = '"' + copilot_unified_email + '"'
```
Two issues here:

(1) Please use the `quot_m` variable here instead of manually inserting quotes. This ensures that the correct type of quotation marks is used.

(2) I am not sure whether the quotation marks are causing any issues in your checks, but both checks seem to be not working. In all the examples, I still see `"""copilot-swe-agentbot"""` in column 13 instead of the expected `"""Copilot"""` in case of commit added events, and `"copilot-swe-agentbot";"""198982749+copilot@users.noreply.github.com"""` in columns 12 and 13 in casae of mentioned or subscribed events.

For the latter, it might be that the `[-1:1]` causes problems for the mentioned/subscribed? Not sure what exactly this does. Could also cause problems for commit_added? You should keep in mind that `event[13]` contains these additional quotation marks while `event[12]` does not. But both checks seem not to work.

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -172,26 +181,44 @@ def is_github_noreply_author(name, email):
+                        # for commit added events, also unify the referenced author in event info 2 if it is a known copilot user
+                        event[13] = '"' + copilot_unified_name + '"'
+                    elif event[8] in (mentioned_event, subscribed_event) and event[12][-1:1] in known_copilot_users_extended:
+                        # for mentioned/subscribed events, also unify the referenced user in event info 1 and 2 if it is a known copilot user
```
It is not only mentioned/subscribed. There could also be other events for which there is user information in columns 12 and 13:

- `assigned`
- `unassigned`
- `review_requested`
- Maybe also for `review_request_removed`? (not sure, this does not occur in the data on which I currently test this).

Could you please check whether there are also other events that take user information in columns 12 and 13?

Please add these events also further down in around line 236 where we filter the GitHub user? And check whether there are other checks for mentioned or subscribed where we need to add the additional events.

bockthom commented on author_postprocessing/author_postprocessing.py:
This needs to handle agents as well.
I suggest to start with agents: If either of the predictions is agents, keep agent.
Then do this for bots and humans as it is now.

bockthom reviewed (changes requested):
I found some time to run the issue analysis on our favorite test project.
This way, I encountered a number of issues that still need to be fixed. I guess the majority of them are easy to fix - only one or two might need a little more thinking on how to fix it. Please see my detailed comments below.

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -164,6 +169,10 @@ def is_github_noreply_author(name, email):
                 if is_github_noreply_author(commit[5], commit[6]):
                     commit[5] = commit[2]
                     commit[6] = commit[3]
+                # unify copilot author if desired
+                if unify_copilot_users and commit[5] in known_copilot_users_extended:
+                    commit[5] = copilot_unified_name
+                    commit[6] = copilot_unified_email
```
This is also not sufficient here:
While the GitHub noreply user can never be an author, it is only replaced in the committer data of a commit. However, the Copilot user can be both, author and committer of a commit.
Therefore, it also needs be replaced accordingly in columns `commit[2]` and `commit[3]`.

Leo-Send committed:
Add copilot user unification for more events

also add agents to bot handling, fix formatting for event_info_2 and
subissues
also fix a typo where strings would not have their quotes correctly
removed

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -172,26 +186,45 @@ def is_github_noreply_author(name, email):
             commit_data = csv_writer.read_from_csv(commit_data_file)
             commit_hash_to_author = {commit[7]: commit[2:4] for commit in commit_data}
-
+            author_name_to_data = {author[1]: author[1:3] for author in author_data_new}
```
I disagree with this suggestion. If `authors.list` does not exist, we should stop here.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -555,21 +670,24 @@ def reformat_events(issue_data):
             if event["event"] == "closed":
                 event["event"] = "state_updated"
                 event["event_info_1"] = "closed"  # new state
-                event["event_info_2"] = "open"  # old state
+                if event["commit"] is not None:
+                    event["event_info_2"] = event["commit"]["hash"]
+                else:
+                    event["event_info_2"] = event["state_reason"]
```
Maybe we can just remove the "old state" in the JIRA data for consistency reasons and let "event_info_2" empty there?

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -758,6 +883,10 @@ def get_user_from_id(idx, buffer_db=user_buffer):
         for event in issue["eventsList"]:
             event["user"] = get_user_from_id(event["user"])
 
+            # for commit_added events, save the commit's author's name in event_info_2
+            if event["event"] == "commit_added":
+                event["event_info_2"] = get_user_from_id(event["event_info_2"])["name"]
```
Technically, @Copilot is right here. But practically, that would increase maintenance load if we would deal with that everywhere. Let's just assume that name collisions don't appear in our data and go with the current implementation.

bockthom commented on bot_processing/bot_processing.py:
```diff
@@ -52,14 +54,15 @@ def run():
     # (the known bots file is the file in which known bots have been added manually and project independent)
     __confdir = os.path.join(args.resdir, os.path.dirname(args.config))
     __known_bots_file = os.path.abspath(os.path.join(__confdir, "known_github_bots.list"))
+    __known_agents_file = os.path.abspath(os.path.join(__confdir, "known_github_agents.list"))
 
     # run processing of bot data:
     # 1) load bot data
     bots = load_bot_data(os.path.join(__srcdir, "bots.csv"), header = True)
     # 2) load user data
     users = load_user_data(os.path.join(__resdir, "usernames.list"))
     # 3) update bot data with user data and additionally add known bots if they occur in the project
-    bots = add_user_data(bots, users, __known_bots_file)
+    bots = add_user_data(bots, users, __known_bots_file, __known_agents_file)
```
Not sure. Do we check if the known bots file exists? What happens if it does not exist?
I suggest to exactly the same for the known agents file: If we do a check for the known bots list, we should also do it for the known agents list. Otherwise I suggest to ignore this comment. Or do you have any other ideas? Technically, one could argue that copilot is a hard-coded known agent, and therefore we should also be able to run this without a separate list. But anyway, I don't want to make it different than for the bots list.

Leo-Send committed:
Add reason for conversation locking

lock reason is saved in event_info_1

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Fix spelling and documentation

docstrings should now more accurately reflect parameters and return
values

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Remove old state from jira state_updated events

For consistency with github events

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Fix jira processing error

previously removed event_info_2 for state_updated event, leading to
crashes of the issue processing. Now, it instead contains an empty
string.
Also fix a minor spelling mistake

Signed-off-by: <s8lesend@stud.uni-saarland.de>

bockthom reviewed (approved):
I didn't spot any errors or inconsistencies in my test run, so let's assume everything works as intended. 
Thanks for your efforts @Leo-Send.

Leo-Send committed:
Fix issue with broken commits in merge events

event_info_1 should remain empty in that case

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Fix 'null' in relatedIssues, subIssues fields crash

These fields are now replaced with empty lists when null

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>
